### ISSUE TITLE
As a content editor, I want to embed external content in editorial and other pages, so that I can include dynamic content in essays.

### ISSUE THREAD

**Issue by rlskoeser** [Princeton-CDH/ppa-django#318]
## notes for testing
When you are editing content in wagtail, you will now have an "embed" option in the blocks you can choose. I've implemented a custom embed finder to embed the visualizations I've created and published as glitch apps (it makes some assumptions, so it won't work for glitch apps that haven't been prepped for embedding). In theory, this will let you embed other kinds of content that support embedding (e.g., youtube videos), although our Content Security Policy configuration may prevent it from being loaded - if you want us to add other urls to the policy, let us know.

You can test embedding with these two glitch project urls:
- https://ppa-collection-stripes2.glitch.me
- https://ppa-collections-treemap.glitch.me/

**Comment by mnaydan** [comment]
On the admin side I can embed the two URLs above in Wagtail. One question I have is whether there should be an option for a caption. The embedded content includes the caption you provide in glitch, but in your essay you have additional captions like "Zoomable treemap diagram of PPA collections, generated with d3.js. Only includes one level of nested collections." There's no way for me to add that caption.

There are a couple problems that occur upon publishing embedded content:

1. **Interactivity** - I couldn't get the treemap diagram to work in Chrome, but then I hard refreshed and now it seems to be working. On your experimental diagram, the stripes exceed the boundaries of the diagram, and the highlighted box when selecting a category is too short (see screenshots A1 and A2). Even weirder, sometimes the stripes diagram is totally blank and non-interactive. I can't figure out when this happens. It's happening in Chrome, SF, FF, even after hard refresh (see A3).
A1: 
<img width="713" alt="Screen Shot 2020-01-03 at 10 23 07 AM" src="https://user-images.githubusercontent.com/22945600/71732051-4fb97800-2e14-11ea-97c7-4a87637b44e5.png">
A2:
<img width="712" alt="Screen Shot 2020-01-03 at 10 30 20 AM" src="https://user-images.githubusercontent.com/22945600/71732064-5a740d00-2e14-11ea-8275-c66a652f3122.png">
A3:
<img width="652" alt="Screen Shot 2020-01-03 at 10 44 11 AM" src="https://user-images.githubusercontent.com/22945600/71733009-8b554180-2e16-11ea-829d-103b12c1018e.png">


2. **Padding** - Text before the embedded content is too close, and there is a large gap after the embedded content before the next text block.
<img width="423" alt="Screen Shot 2020-01-03 at 10 33 52 AM" src="https://user-images.githubusercontent.com/22945600/71732220-ba6ab380-2e14-11ea-8af4-a59e01cf407e.png">

FWIW, Youtube videos can be embedded and play (AV works) but have the same padding problems and are too small.
<img width="646" alt="Screen Shot 2020-01-03 at 10 42 35 AM" src="https://user-images.githubusercontent.com/22945600/71733109-c8b9cf00-2e16-11ea-8ff9-c4729951cc42.png">

3. I don't know if this would ever actually be a problem, but I tested putting the same URL in a couple of different places in the essay, and the placement of the content on the published page does not correspond the their placement in Wagtail. The two treemap diagrams are appearing right next to each other even though I didn't put them there. The caption for one of the treemap diagrams (missing on the screenshot) appears at the end of the essay, where the diagram is supposed to be.
<img width="431" alt="Screen Shot 2020-01-03 at 11 03 27 AM" src="https://user-images.githubusercontent.com/22945600/71733963-cbb5bf00-2e18-11ea-9c81-a018abe991df.png">

**Comment by rlskoeser** [comment]
The wagtail embed block I'm using is a wagtail default, which makes me think the lack of caption is standard - although I guess the wagtail default image embed didn't include captions and we wanted them there.  For this content, I'm fine with editing the text in my glitch apps to include appropriate captions.  If you were embedding other content (e.g., a video or something)* do you think you'd want to caption it? Would you _need_ to caption it?  

I think the glitch app not loading is related to our Content Security Policy (CSP) configuration - I'll take another look at that. I'd previously assumed that the layout errors you noticed in the collection stripes app were something I needed to correct independent of embedding it - but it occurs to me now that maybe there are some style conflicts causing it when it is embedded. 

I see the spacing errors on the test site now - weird. I didn't pay that much attention to the styles for the embed block when I first implemented; I'll look at those now and see if I can make them consistent with the image spacing styles.

The duplicate treemap is funny! Took me a moment, but I think that's because of the way the javascript is written - I think it's adding the same SVG twice to the same html container, which is why you get the text where you embedded the second one but not the diagram.

* FWIW, I don't think our current Content Security Policy will allow you to embed content from other sites. But we can revise it as needed.

**Comment by rlskoeser** [comment]
@mnaydan I've revised the styles for embedded content some (wagtail was making some weird assumptions I had to override), and also made some revisions to my glitch apps to address some of the things you identified.  Among other things, I added a version of the PPA caption style that I can use for in my embedded code by adding the appropriate CSS class, and I think it looks nicer with that. 

I think the problem with the collection stripes not displaying is actually a conflict with the treemap, because when I test a page with just the collections it works fine (see my "embed test two" on the test site). The two visualizations are written with different versions of d3 which are not compatible (the treemap was adapted from existing code). That's not specific to the embed logic, but it is something I'm glad you found and that I'll have to work on.

**Comment by mnaydan** [comment]
I retested this across browsers and devices and it looks good. I tried adding the embedded content again on an "embed test three" editorial page and also checked the pages I had been testing with before. The padding and formatting is fixed now. If you are okay with adding your own captions in glitch, I don't think we need the functionality to add a caption. 

The interactivity and style problems with the experimental diagram are still present but as we discussed those need to be fixed in glitch independently of this feature.
